File: tasks/text-redact/javascript/environment/src/redact.js
"use strict";

/** Scan for the literal one position at a time. */
function findSpans(rule, text) {
  const spans = [];
  const width = rule.value.length;
  let index = 0;
  while (index + width <= text.length) {
    if (text.slice(index, index + width) === rule.value) {
      spans.push([index, index + width]);
    }
    index += 1;
  }
  return spans;
}

/** Spans are reported in the order they were found. */
function mergeSpans(spans) {
  return spans.map(([start, end, id]) => ({ start, end, rules: [id] }));
}

function applyMask(text, spans, mask) {
  const characters = text.split("");
  for (const span of spans) {
    for (let index = span.start; index < span.end; index += 1) {
      characters[index] = mask;
    }
  }
  return characters.join("");
}

module.exports = { applyMask, findSpans, mergeSpans };

File: tasks/text-redact/python/environment/src/redact.py
"""Find the regions a rule matches and mask them."""


def find_spans(rule, text):
    """Scan for the literal one position at a time."""
    spans = []
    width = len(rule["value"])
    index = 0
    while index + width <= len(text):
        if text[index : index + width] == rule["value"]:
            spans.append((index, index + width))
        index += 1
    return spans


def merge_spans(spans):
    """Spans are reported in the order they were found."""
    return [
        {"start": start, "end": end, "rules": [identifier]}
        for start, end, identifier in spans
    ]


def apply_mask(text, spans, mask):
    characters = list(text)
    for span in spans:
        for index in range(span["start"], span["end"]):
            characters[index] = mask
    return "".join(characters)
